src/bytes_mut.rs

```diff
@@ -1246,24 +1092,70 @@ impl Shared {
-        self.ref_count.load(Ordering::Acquire) == 1
+        self.refcount.load(Ordering::Acquire) == 1
     }
-}
 
-fn original_capacity_to_repr(cap: usize) -> usize {
-    let width = PTR_WIDTH - ((cap >> MIN_ORIGINAL_CAPACITY_WIDTH).leading_zeros() as usize);
-    cmp::min(
-        width,
-        MAX_ORIGINAL_CAPACITY_WIDTH - MIN_ORIGINAL_CAPACITY_WIDTH,
-    )
-}
+    fn layout_for_size(size: usize) -> Layout {
+        let layout = Layout::new::<Shared>();
+        let total_size = layout
+            .size()
+            .checked_add(size)
+            .expect("Overflow during layout calculation");
+        let combined_layout =
+            Layout::from_size_align(total_size, layout.align()).expect("Error calculating layout");
+        combined_layout
+    }
+
+    unsafe fn increment_refcount(self_ptr: *mut Self) {
+        debug_assert!(!self_ptr.is_null());
+        let old_size = (*self_ptr).refcount.fetch_add(1, Ordering::Relaxed);
+
+        if old_size > isize::MAX as usize {
+            crate::abort();
+        }
+    }
+
+    unsafe fn release_refcount(self_ptr: *mut Self) {
+        if self_ptr.is_null() {
+            return;
+        }
 
-fn original_capacity_from_repr(repr: usize) -> usize {
-    if repr == 0 {
-        return 0;
+        if (*self_ptr).refcount.fetch_sub(1, Ordering::Release) != 1 {
+            return;
+        }
+
+        // The shared state should be released
+
+        // This fence is needed to prevent reordering of use of the data and
+        // deletion of the data.  Because it is marked `Release`, the decreasing
+        // of the reference count synchronizes with this `Acquire` fence. This
+        // means that use of the data happens before decreasing the reference
+        // count, which happens before this fence, which happens before the
+        // deletion of the data.
+        //
+        // As explained in the [Boost documentation][1],
+        //
+        // > It is important to enforce any possible access to the object in one
+        // > thread (through an existing reference) to *happen before* deleting
+        // > the object in a different thread. This is achieved by a "release"
+        // > operation after dropping a reference (any access to the object
+        // > through this reference must obviously happened before), and an
+        // > "acquire" operation before deleting the object.
+        //
+        // [1]: (www.boost.org/doc/libs/1_55_0/doc/html/atomic/usage_examples.html)
+        atomic::fence(Ordering::Acquire);
+
+        alloc::alloc::dealloc(
+            self_ptr as *mut u8,
+            Self::layout_for_size((*self_ptr).capacity),
+        )
     }
 
-    1 << (repr + (MIN_ORIGINAL_CAPACITY_WIDTH - 1))
+    unsafe fn data_ptr_mut(&mut self) -> NonNull<u8> {
+        let mut end_addr = self as *const Shared as usize;
+        end_addr += mem::size_of::<Shared>();
+        NonNull::new_unchecked(end_addr as *mut u8)
+    }
```

Comment:
nit: i feel like it's a bit cleaner to use `offset`:

```suggestion
    unsafe fn data_ptr_mut(&mut self) -> NonNull<u8> {
        let shared_addr = self as *mut Self;
        let data_addr = shared_addr.offset(1) as *mut u8;
        NonNull::new_unchecked(data_addr)
    }
```

Reply:
Maybe. It should lead to the same result, but I'm not sure whether I should trust Rust more or less about casting the pointer at the end of `Shared` to a `u8` than at the beginning :)

Reply:
They are the same result, for sure. I guess I was more concerned about the number of conversions that occur in the previous version, especially the `&mut` -> `*const` -> `*mut`:

`&mut Shared` -> `*const Shared` -> `usize` `*mut u8`

vs

`&mut Shared` -> `*mut Shared` -> `*mut u8`

Reply:
Don't cast a `&mut T` to a `*const T`. It involves first coercing it to an `&T`, which means that it would be unsound to perform any writes through the resulting pointer. Try running [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=da1be247e82bc90ae919153b9c9e05ee) using miri from the tools.